### ISSUE TITLE
Optimize slow bottleneck tests

### ISSUE THREAD

**Issue by tiborvass** [moby/moby#19425]
The time the integration test suite takes to finish is getting out of control.
We would like to ask your help in optimizing some of the longest test cases.
This is another of those "non-pleasant" tasks of which, if we all do just one, we can get it all done extra fast.

If you would like to help out, please comment on this issue with the name of the test you would like to take care of; we'll add your Github handle next to the test name so that others can tell which tests still need an owner.

**UPDATE**: optimizing a test consists of making sure the test code is reasonably efficient. However if the slowdown is a sign of not efficient code in Docker itself, then that would need to be addressed too.

I'm shamelessly stealing the following from @estesp's gist https://gist.github.com/estesp/40a056a8262eafb62dd1:
#### Sampled from a 1/12/2016 run of standard Linux x64 engine _janky_ run.

Total time of these 25 tests adds up to 663.8 seconds

These 25 make up 2.2% of total tests, but ~29% of total test runtime.

```
DockerDaemonSuite.TestDaemonNoSpaceleftOnDeviceError            107.143s DockerTrustSuite.TestTrustedPushWithFailingServer               95.405s
DockerSuite.TestRunSeccompDefaultProfile                        81.711s
DockerSuite.TestEventsImagePull                                 43.321s
DockerSuite.TestBuildEnvUsage2                                  32.791s
DockerSuite.TestDaemonStartWithDaemonCommand                    25.648s
DockerHubPullSuite.TestPullAllTagsFromCentralRegistry           24.504s
DockerSuite.TestEventsOOMDisableTrue                            20.288s
DockerHubPullSuite.TestPullFromCentralRegistryImplicitRefParts  19.389s
DockerSuite.TestBuildHistory                                    18.234s
DockerSuite.TestBuildUsersAndGroups                             16.550s
DockerSuite.TestRunUnshareProc                                  16.196s
DockerSuite.TestVolumeCliCreate                                 15.118s
DockerSuite.TestApiStatsNetworkStatsVersioning                  14.699s
DockerSuite.TestBuildDockerignoringWildDirs                     14.443s
DockerSuite.TestApiStatsNetworkStats                            13.200s
DockerSuite.TestBuildSixtySteps                                 12.663s
DockerHubPullSuite.TestPullNonExistingImage                     12.259s
DockerDaemonSuite.TestDaemonRestartWithContainerRunning         12.110s
DockerDaemonSuite.TestDaemonCorruptedLogDriverAddress           12.007s
DockerSuite.TestDaemonIPv6FixedCIDR                             11.564s
DockerSuite.TestInfoDiscoveryInvalidAdvertise                   11.504s
DockerSuite.TestDaemonIPv6FixedCIDRAndMac                       11.451s
DockerSuite.TestHelpTextVerify                                  11.039s
DockerSuite.TestEventsAttach                                    10.635s
```

Additional tests (suggested by @aaronlehmann)

```
DockerSuite.TestEventsContainerFailStartDie 40.018s
DockerSuite.TestRunSeccompAllowPrivCloneUserns  40.302s
DockerSuite.TestRunSeccompProfileDenyCloneUserns        45.066s
DockerSuite.TestRunSeccompUnconfinedCloneUserns 40.056s
```

And more (suggested by @wenchma)

```
DockerNetworkSuite.TestDockerNetworkInternalMode   11.268s
```

TODO:
- [x] DockerDaemonSuite.TestDaemonNoSpaceleftOnDeviceError (#19551 @aaronlehmann)
- [x] DockerTrustSuite.TestTrustedPushWithFailingServer (#23217 @clnperez)
- [x] DockerSuite.TestRunSeccompDefaultProfile (#19725 @WeiZhang555)
- [x] DockerSuite.TestEventsImagePull (test is now 2.273s)
- [x] DockerSuite.TestBuildEnvUsage2 (#19600 @aaronlehmann)
- [x] DockerSuite.TestDaemonStartWithDaemonCommand (#20476 @wenchma)
- [ ] DockerHubPullSuite.TestPullAllTagsFromCentralRegistry
- [x] DockerSuite.TestEventsOOMDisableTrue (#19619 @WeiZhang555)
- [ ] DockerHubPullSuite.TestPullFromCentralRegistryImplicitRefParts
- [x] DockerSuite.TestBuildHistory (#20880 @yongtang)
- [x] DockerSuite.TestBuildUsersAndGroups (#21056 @askb)
- [x] DockerSuite.TestRunUnshareProc (#20965 @yongtang)
- [ ] DockerSuite.TestVolumeCliCreate (@askb)
- [x] DockerSuite.TestApiStatsNetworkStatsVersioning (#19800 @wenchma)
- [x] DockerSuite.TestBuildDockerignoringWildDirs (#20929 @yongtang)
- [x] DockerSuite.TestApiStatsNetworkStats (#19800 @wenchma)
- [ ] DockerSuite.TestBuildSixtySteps
- [ ] DockerHubPullSuite.TestPullNonExistingImage
- [x] DockerDaemonSuite.TestDaemonRestartWithContainerRunning (#19663 @mountkin)
- [x] DockerDaemonSuite.TestDaemonCorruptedLogDriverAddress (#19708 @wenchma)
- [x] DockerSuite.TestDaemonIPv6FixedCIDR (#19658 @wenchma)
- [ ] DockerSuite.TestInfoDiscoveryInvalidAdvertise
- [x] DockerSuite.TestDaemonIPv6FixedCIDRAndMac (#19658 @wenchma)
- [ ] DockerSuite.TestHelpTextVerify (@sainath14)
- [x] DockerSuite.TestEventsAttach (#19703 @mountkin)
- [x] DockerSuite.TestEventsContainerFailStartDie (#19713 @WeiZhang555)
- [ ] DockerSuite.TestRunSeccompAllowPrivCloneUserns
- [ ] DockerSuite.TestRunSeccompProfileDenyCloneUserns
- [ ] DockerSuite.TestRunSeccompUnconfinedCloneUserns
- [x] DockerNetworkSuite.TestDockerNetworkInternalMode (#20903 @wenchma )

**Comment by crosbymichael** [comment]
![yes](http://gifs.joelglovier.com/swoon/ohhh-minion.gif)

**Comment by HackToday** [comment]
hi @tiborvass  I'like to offer help, but what's the target date for those test fixing ?

Also one question, if those tests coverage are fine. it seems can not reduce those test time, if bottle neck is related with docker code, do we need address that ?

**Comment by aditirajagopal** [comment]
@jecarey and I would be interested in helping out

**Comment by thaJeztah** [comment]
Added "flakytest" label; although not entirely accurate, it makes it easier to track :)

**Comment by StefanScherer** [comment]
There also is a kind of test that takes very long. The helper `.ensure-syscall-test` takes 6-8 minutes on ARM for just compiling "three little C files".

```
20:57:15 ---> Making bundle: .ensure-syscall-test (in bundles/1.10.0-dev/test-integration-cli)
21:03:30 + go test -test.timeout=180m -check.v github.com/docker/docker/integration-cli
```

and eg. on Janky (from https://jenkins.dockerproject.org/job/Docker-PRs/22578/console) it takes over 5 minutes:

```
16:55:10 ---> Making bundle: .ensure-syscall-test (in bundles/1.10.0-dev/test-integration-cli)
17:00:33 + go test -test.timeout=120m -check.v github.com/docker/docker/integration-cli
```

**Comment by LK4D4** [comment]
@StefanScherer Yup, very long :/

**Comment by MHBauer** [comment]
I took a shot at DockerDaemonSuite.TestDaemonNoSpaceleftOnDeviceError in #19484.

If my guess is right, we need to try and get rid of pulls that talk to the remote hub. I can see pull in the title of maybe a quarter of the tests, and I'd bet that some of the other tests have a pull in them, or maybe a push, but probably some communication with remote docker hub registry.

**Comment by unclejack** [comment]
@StefanScherer There's no surprise when it comes to how long it takes. It's pulling an image from the registry.

**Comment by StefanScherer** [comment]
@unclejack Yes, but the debian/jessie should already be "downloaded" through the `.ensure-frozen-images`. What takes time I think is building the docker image with the Dockerfile installing the build essentials etc.

**Comment by unclejack** [comment]
That image is quite big. The integration-cli tests end up using the vfs graph driver. Using the vfs graph driver means every single layer will contain all the data of all layers below it because there's no copy on write for this graph driver. This means using huge images with lots of files will slow down the tests a lot.

Is there a reason behind using such a big image to compile something we could also compile in the main Docker container and copy the resulting binaries to a smaller container? We should get rid of this huge image.

**Comment by aaronlehmann** [comment]
`DockerDaemonSuite.TestDaemonNoSpaceleftOnDeviceError` will be fixed by https://github.com/docker/docker/pull/19551

**Comment by aaronlehmann** [comment]
ping @tianon

**Comment by aaronlehmann** [comment]
I'll nominate the following additional tests (taken from a recent janky build):

```
DockerSuite.TestEventsContainerFailStartDie	40.018s
DockerSuite.TestRunSeccompAllowPrivCloneUserns  40.302s
DockerSuite.TestRunSeccompProfileDenyCloneUserns        45.066s
DockerSuite.TestRunSeccompUnconfinedCloneUserns 40.056s
```

**Comment by thaJeztah** [comment]
Thanks @aaronlehmann closed that one test, and added the other ones at the end

**Comment by tiborvass** [comment]
@HackToday @aditirajagopal @jecarey sorry i somehow missed those pings.

Yes by optimizing the tests I also include optimizing Docker code if needed. Or modifying the Dockerfile or whatever it takes. I updated the description of the issue to reflect this.

Like I said, if you want to tackle one of the tests, just write the name of your test so that we can put your name next to it.

**Comment by aaronlehmann** [comment]
Submitted a marginal improvement to `TestBuildEnvUsage2` in https://github.com/docker/docker/pull/19600

**Comment by WeiZhang555** [comment]
I'll take care of `TestEventsOOMDisableTrue`, already pushed a PR #19619 
@tiborvass

**Comment by thaJeztah** [comment]
Awesome @wenchma, I added it to the list!

**Comment by wenchma** [comment]
I will fix DockerSuite.TestDaemonIPv6FixedCIDR and DockerSuite.TestInfoDiscoveryInvalidAdvertise in docker_cli_daemon_test.go

**Comment by wenchma** [comment]
@tiborvass #19658 is to fix TestDaemonIPv6FixedCIDR and  `TestDaemonIPv6FixedCIDRAndMac`
sorry for wrong copy `DockerSuite.TestInfoDiscoveryInvalidAdvertise`, but it is also ok for me to fix `TestInfoDiscoveryInvalidAdvertise` ,  :smile:

**Comment by wenchma** [comment]
I will fix `DockerDaemonSuite.TestDaemonCorruptedLogDriverAddress`

**Comment by WeiZhang555** [comment]
Fix for `TestEventsContainerFailStartDie` is already submitted, PR #19713

**Comment by cyphar** [comment]
Just had this:

```
14:15:46 PASS: docker_cli_run_unix_test.go:846: DockerSuite.TestRunSeccompAllowPrivCloneUserns	43.873s
14:20:49 PASS: docker_cli_run_unix_test.go:867: DockerSuite.TestRunSeccompDefaultProfile	295.090s
14:20:51 PASS: docker_cli_run_unix_test.go:756: DockerSuite.TestRunSeccompProfileDenyChmod	1.965s
14:21:39 PASS: docker_cli_run_unix_test.go:822: DockerSuite.TestRunSeccompProfileDenyCloneUserns	46.629s
14:21:49 PASS: docker_cli_run_unix_test.go:728: DockerSuite.TestRunSeccompProfileDenyUnshare	8.848s
14:21:52 PASS: docker_cli_run_unix_test.go:785: DockerSuite.TestRunSeccompProfileDenyUnshareUserns	3.000s
```

**Comment by WeiZhang555** [comment]
Trying to improve `TestRunSeccompDefaultProfile` in PR #19725

**Comment by wenchma** [comment]
I will fix the `DockerSuite.TestApiStatsNetworkStatsVersioning`  and `DockerSuite.TestApiStatsNetworkStats` .

**Comment by sillyousu** [comment]
Speed up `DockerSuite.TestVolumeCliCreate`  with PR https://github.com/docker/docker/pull/19970

**Comment by wenchma** [comment]
Fix `DockerSuite.TestDaemonStartWithDaemonCommand`
related PR: #20476

**Comment by thaJeztah** [comment]
Thanks @wenchma - added

**Comment by sainath14** [comment]
TestRunOOMExitCode - 401.719
TestInspectOomKilledTrue - 402.732

Anyone seeing this on a 1.11 version?

**Comment by yongtang** [comment]
I create a pull request to try to speed up DockerSuite.TestBuildHistory:
https://github.com/docker/docker/pull/20880

On an AWS EC2 t2.large (Ubuntu 14.04) instance the execution time improved from 13s to about 2s:
```
Before:
PASS: docker_cli_history_test.go:15: DockerSuite.TestBuildHistory    13.067s
After:
PASS: docker_cli_history_test.go:15: DockerSuite.TestBuildHistory    2.672s
```
Let me know if there are any issues.
cc @tiborvass @thaJeztah

**Comment by thaJeztah** [comment]
Thanks @yongtang, I edited the lists!

**Comment by wenchma** [comment]
Just nominate another test from recent `experiment` check:

    11:58:08 PASS: docker_cli_network_unix_test.go:1423: DockerNetworkSuite.TestDockerNetworkInternalMode	11.268s

and provide an enhancement: #20903

**Comment by yongtang** [comment]
I submit another pull request https://github.com/docker/docker/pull/20914 to try to speed
up DockerHubPullSuite.TestPullNonExistingImage.
As the execution time was mostly due to the fact that ```docker pull``` of
non-existing images has been done multiple (10) times, this pull request
tries to run ```docker pull``` in parallel.

On an AWS EC2 t2.large (Ubuntu 14.04) instance the execution time improved from 10s to about 1.5s:
```
Before:
PASS: docker_cli_pull_test.go:40: DockerHubPullSuite.TestPullNonExistingImage	10.451s
After:
PASS: docker_cli_pull_test.go:41: DockerHubPullSuite.TestPullNonExistingImage	1.552s
```
Let me know if there are any comments or concerns.
cc @tiborvass @thaJeztah

**Comment by yongtang** [comment]
I also take a look at DockerSuite.TestEventsImagePull. It seems that at the moment
this test runs pretty fast:

```PASS: docker_cli_events_test.go:209: DockerSuite.TestEventsImagePull    2.257s```

I think TestEventsImagePull probably is not a major issue right now.

**Comment by yongtang** [comment]
I submitted a pull request https://github.com/docker/docker/pull/20929 to try to speed up TestBuildDockerignoringWildDirs.

The long execution time in this test was mostly due to the fact that there are many (19) RUN instructions inside the Dockerfile. This PR collapses those RUN instructions into one and speed up the execution time significantly. The purpose of this test (ignoring the wild dirs in docker build) has not been changed.

On an AWS EC2 t2.large (Ubuntu 14.04) instance the execution time improved from 10s to about 1.4s:
```
Before:
PASS: docker_cli_build_test.go:3690: DockerSuite.TestBuildDockerignoringWildDirs
10.976s
After:
PASS: docker_cli_build_test.go:3690: DockerSuite.TestBuildDockerignoringWildDirs
1.455s
```
Let me know if there are any comments or concerns.

**Comment by Fredjet** [comment]
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  fredjet15@gmail.com
    Domain ventureacceleration.com has exceeded the max emails per hour (57/50 (113%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from github-smtp2-ext6.iad.github.net ([192.30.252.197]:43945 helo=github-smtp2b-ext-cp1-prd.iad.github.net)
	by eaglepeak.ovasp.com with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.86_1)
	(envelope-from <noreply@github.com>)
	id 1abd1X-0000af-2E
	for fred@ventureacceleration.com; Thu, 03 Mar 2016 18:51:15 -0500
Date: Thu, 03 Mar 2016 15:50:34 -0800
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1457049034;
	bh=7COboAx8pqTeT/6DTilmts7GAl2XTFQkscz5+4Gyhg0=;
	h=From:Reply-To:To:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=lxT/HhCYsoKeY/gahwpIWFNODDRQ4ToD3iJfbLiSVnkq5/JF0MGmlCJmF1qNOjRNN
	 RiazVTIMe7EH5KUr0kwLHUSFzSruIpQCUmpbHL+4oTMYQRootVwpI7+H9dDkXAqmfh
	 iIrTdgWEzGz8jm3fjb/ptFNfuHM4wDeOIRviIYdk=

**Comment by yongtang** [comment]
Another pull request https://github.com/docker/docker/pull/20929 is to speed up TestRunUnshareProc.

This fix tries to execute the 'docker run' command in parallel in order to speed up the execution of the overall test. As there are only three 'docker run' commands in this test, the overall improvement is only moderate.

On an AWS EC2 t2.large (Ubuntu 14.04) instance the execution time improved from 10s to about 4.4s:
```
Before:
PASS: docker_cli_run_test.go:2990: DockerSuite.TestRunUnshareProc	10.906s
After:
PASS: docker_cli_run_test.go:2990: DockerSuite.TestRunUnshareProc	4.444s
```
Let me know if there are any comments or concerns.

**Comment by aaronlehmann** [comment]
`TestRunUnshareProc` PR mentioned above is https://github.com/docker/docker/pull/20965

**Comment by yongtang** [comment]
Thanks @aaronlehmann! Copied the wrong link.

**Comment by askb** [comment]
@tiborvass I would be happy to help out here!

**Comment by thaJeztah** [comment]
@askb cool! If you have a test in mind that you want to work on, leave a comment, then I'll put the PR behind it in the description

**Comment by askb** [comment]
@thaJeztah will take `DockerSuite.TestVolumeCliCreate`

**Comment by thaJeztah** [comment]
Thanks! Added your name :+1:

**Comment by sainath14** [comment]
I would like to contribute here.  Can I work on optimizing DockerSuite.TestHelpTextVerify?

**Comment by thaJeztah** [comment]
@sainath14 of course! added your name

**Comment by aaronlehmann** [comment]
Reopening - looks like this was closed accidentally because of GitHub's interpretation of the summary in #21056.

**Comment by askb** [comment]
@aaronlehmann thanks for re-opening the pull request.

@sillyousu Sorry, I dint notice you are working on this tests `DockerSuite.TestVolumeCliCreate`. I could look at other tests. This needs to addressed in the docker code, as most of the time actually spent between sleeps/retries, to check if plugin is loaded in func `loadWithRetry()`. 

From the previous pull request and other PR's, its seems favorouble doing retries is this scenario which would otherwise cause issues during daemon startup. Here are some thoughts after spending some time digging in the code. The first part of the test takes a large time (-ve case), for the value it adds, would it ideal to split the test `DockerSuite.TestVolumeCliCreate` into separate cases (+ve case which works as expected), and optinally exclude the part of the test. Alternatively, would it be good to introduce a new flag (or hook) during daemon startup, which sets `retry = false` useful only for running the tests. Other suggestions are welcome!

**Comment by yongtang** [comment]
I take a further look at this issue, it seems that there are some changes recently that caused certain tests to run even longer. For example, from one janky (03/14/2016) test:

https://jenkins.dockerproject.org/job/Docker-PRs/25108/console

we could see that the following 5 tests exceed 30s. (There might be some regressions compared with the result from 01/12/2016. For example, TestDockerNetworkInternalMode increased from 11s to 47s):
```
PASS: docker_cli_run_unix_test.go:852: DockerSuite.TestRunSeccompDefaultProfile	31.484s
PASS: docker_cli_pull_test.go:192: DockerHubPullSuite.TestPullAllTagsFromCentralRegistry	31.530s
PASS: docker_cli_restart_test.go:145: DockerSuite.TestRestartWithPolicyUserDefinedNetwork	33.662s
PASS: docker_cli_push_test.go:330: DockerTrustSuite.TestTrustedPushWithFailingServer	35.379s
PASS: docker_cli_network_unix_test.go:1423: DockerSuite.TestDockerNetworkInternalMode	47.961s
```

Of those 5 tests, `TestDockerNetworkInternalMode` and `TestRestartWithPolicyUserDefinedNetwork` are mainly caused by the long time for the `exec` ping test cases.

I created two pull requests, https://github.com/docker/docker/pull/21200 and https://github.com/docker/docker/pull/21201, which addresses the above mentioned two tests by using coroutines to parallelize the test cases. There are some pretty good improvements with the fixes:
```
BEFORE:
PASS: docker_cli_network_unix_test.go:1423: DockerSuite.TestDockerNetworkInternalMode	47.961s
AFTER:
PASS: docker_cli_network_unix_test.go:1423: DockerSuite.TestDockerNetworkInternalMode	32.718s
```

```
BEFORE:
PASS: docker_cli_restart_test.go:145: DockerSuite.TestRestartWithPolicyUserDefinedNetwork	33.662s
AFTER:
PASS: docker_cli_restart_test.go:146: DockerSuite.TestRestartWithPolicyUserDefinedNetwork	18.287s
```

**Comment by clnperez** [comment]
Looks like TestRunSeccompDefaultProfile was split into two (still ~30s combined) tests: 
TestRunSeccompDefaultProfileNS
TestRunSeccompDefaultProfileAcct

I just ran DockerSuite.TestRunSeccompAllowPrivCloneUserns  and it took 3.901s, so we can prob take that out out.

**Comment by clnperez** [comment]
@tiborvass PR #23217 should fix TestTrustedPushWithFailingServer.

**Comment by dnephin** [comment]
@estesp do you have a script to generate the report that is linked in the description of this issue?

**Comment by estesp** [comment]
@dnephin sadly it appears I do not. I remember pulling the raw test run log output from jenkins and the rest was most likely some hacky awk/sed + sort + simple math in bash.

**Comment by dnephin** [comment]
Ok, I might look at recreating something so that we can run it periodically.  I've been trying to get a profile of the test suite using https://gist.github.com/dnephin/3509caeb8ac7066e73e2155e716b4927 but the profiles I'm getting out are empty for some reason. I must be missing something about the golang profiler. If anyone has had success getting a profile from the integration test suite, or knows what I'm doing wrong, please let me know.

**Comment by dnephin** [comment]
Ok, I wrote up a script and put it into a [docker image](https://hub.docker.com/r/dnephin/docker-testsuite-report/)

````
docker run --rm dnephin/docker-testsuite-report 31507
````


```
Elapsed: 1:07:31
Elapsed integration suite: 0:50:21
Sum test times: 0:32:09
Time to first integration test: 0:10:02 
Time after last integration test: 0:07:08
Test count: 1429

Slowest tests:
   PASS: docker_cli_push_test.go:666: DockerRegistryAuthTokenSuite.TestPushMisconfiguredTokenServiceResponseError	50.144s
   PASS: docker_cli_run_test.go:3105: DockerSuite.TestRunUnshareProc	43.859s
   PASS: docker_cli_pull_test.go:201: DockerHubPullSuite.TestPullAllTagsFromCentralRegistry	33.525s
   PASS: docker_cli_run_unix_test.go:1105: DockerSuite.TestRunSeccompDefaultProfileNS	31.771s
   PASS: docker_cli_network_unix_test.go:1621: DockerSuite.TestDockerNetworkInternalMode	30.940s
   PASS: docker_cli_network_unix_test.go:958: DockerNetworkSuite.TestDockerNetworkDriverUngracefulRestart	27.392s
   PASS: docker_api_swarm_test.go:624: DockerSwarmSuite.TestApiSwarmRaftQuorum	25.981s
   PASS: docker_cli_run_unix_test.go:1076: DockerSuite.TestRunSeccompDefaultProfileAcct	25.896s
   PASS: docker_api_swarm_test.go:694: DockerSwarmSuite.TestApiSwarmNodeRemove	24.595s
   PASS: docker_api_swarm_test.go:802: DockerSwarmSuite.TestApiSwarmLeaveOnPendingJoin	22.170s
```

Something that I found interested is this:
```
Elapsed integration suite: 0:50:21
Sum test times: 0:32:09
```

We spend over 1/3 of the test time in setup/teardown fixtures.

**Comment by aaronlehmann** [comment]
> PASS: docker_cli_push_test.go:666: DockerRegistryAuthTokenSuite.TestPushMisconfiguredTokenServiceResponseError   50.144s

ping @dmcgowan

**Comment by samwagg** [comment]
I was taking a look at `DockerHubPullSuite.TestPullAllTagsFromCentralRegistry`, which apparently takes a long time because it is verifying --all-tags functionality using busybox. I was wondering, would it make sense to host a dedicated dummy repo for this test that has say, 2-3 tiny images? I'm sure there are pre-existing repos with fewer tags/images, but it seems to me it would be nice to have a long-term guarantee that the repo used in this test will have favorable characteristics. 

Please let me know if I've said anything dumb, I'm just starting my journey of understanding the project and contributing :)

**Comment by AkihiroSuda** [comment]
Linking my PR https://github.com/docker/docker/pull/26684 which parallelizes the tests (49 min --> 13 min, IMO it can be shorten to less than about 5 min in future optimized version) to this issue.
Anyone please look into my PR?

**Comment by cyphar** [comment]
> I was wondering, would it make sense to host a dedicated dummy repo for this test that has say, 2-3 tiny images?

Note that some people run the test suite against their own registry (to test their packages as a distribution), so this would make the test not usable for those people.

**Comment by aaronlehmann** [comment]
> Note that some people run the test suite against their own registry (to test their packages as a distribution), so this would make the test not usable for those people.

For those people, why does it matter if this test pulls busybox or some other repo?

I'm +1 on switching to a different repo that takes less time to pull.

**Comment by samwagg** [comment]
> I'm +1 on switching to a different repo that takes less time to pull.

My specific suggestion was to add a dedicated repo that will remain appropriate for this test long-term (small number of small images, but more than 1).  Any chance an existing namespace or registry exists for images used internally (neither official nor personal seem appropriate for this)? Would it be better to just switch to a repo that is currently small and move on (not sure this is worth much effort)? I appreciate the guidance!

In any case, please add my name to `DockerHubPullSuite.TestPullAllTagsFromCentralRegistry` if that's okay.

**Comment by cyphar** [comment]
@aaronlehmann Oh, I misread what the point was. I thought that @samwagg was suggesting that we run it on a dummy registry (not a dummy image). Never mind, I hadn't had much sleep while responding. :wink:
